design/design-doc-composition-functions.md

```diff
@@ -175,32 +179,557 @@ metadata:
+  rsp := function.NewResponseTo(req)
+  function.Normalf(rsp, "successfully read desired composite resource %q", xr.GetName())
+
+  return rsp, nil
+}
+```
+
+To write a Function, replace the boilerplate with your own logic. Here’s an
+example that uses a mockup of a Go SDK to achieve the same goal as lab 4 from
+[the Kubecon EU ContribFest][kubecon-eu-contribfest].
+
+```yaml
+apiVersion: contribfest.crossplane.io/v1alpha1
+kind: XRobotGroup
+metadata:
+  name: somename
+spec:
+  count: 5
+```
+
+This example takes an `XRobotGroup` XR and uses its `spec.count` to compose the
+desired number of Robot resources:
+
+```go
+package main
+
+import (
+    "context"
+    "fmt"
+    "math/rand"
+
+    "github.com/upbound/provider-dummy/apis/iam/v1alpha1"
+
+    function "github.com/crossplane/fn-sdk-go"
+    "github.com/crossplane/fn-sdk-go/pb"
+    "github.com/crossplane/fn-sdk-go/errors"
+)
+
+var colors = []string{"red", "green", "blue", "yellow", "orange"}
+
+// Function defines the logic of your Function.
+func Function(ctx context.Context, req *pb.RunFunctionRequest) (*pb.RunFunctionResponse, error) {
+  // Get the desired XR from the request.
+  xr := function.NewEmptyCompositeResource()
+  function.GetCompositeResource(req, xr)
+
+  // Get the desired resource count from the XR.
+  count, err := xr.GetInteger("spec.count")
+  if err != nil {
+      return errors.Wrap(err, "could not get desired resource count")
+  }
+
+  // Create a response to return. This deep-copies any existing desired state.
+  rsp := function.NewResponseTo(req)
+
+  // Ensure the desired number of robots exist.
+  for i := 0; i < int(count); i++ {
+      name := fmt.Sprintf("robot-%d")
+
+      // Get the desired composed resource (if any) from the request.
+      // Since we're using Go we can import and work with API types if we want.
+      robot := &v1alpha1.Robot{}
+      function.GetComposedResource(req, name, robot)
+
+      // The robot's external name should be derived from the XR's.
+      function.SetExternalName(robot, fmt.Sprintf("%s-%s", function.GetExternalName(xr), name))
+
+      // Give this robot a random color!
+      if robot.Spec.ForProvider.Color != "" {
+          c := colors[rant.Intn(len(colors))]
+          robot.Spec.ForProvider.Color = c
+
+          // Add a result indicating that we set the color. Crossplane will
+          // surface this as a Kubernetes event and a debug log.
+          function.Normalf(rsp, "set robot %q color to %q", name, c)
+      }
+
+      // Set our new desired robot state. This will be a no-op if our robot
+      // already existed.
+      function.SetComposedResource(rsp, name, robot)
+  }
+
+  return rsp, nil
+}
+```
+
+The goals of a Composition Function SDK are to:
+
+* Expose an API that makes writing Functions intuitive and self-documenting.
+* Steer Function authors toward best practices, and away from anti-patterns.
+* Eliminate boilerplate, like setting up a gRPC server.
```

Comment:
For the server setup, I think we can get rid of basically everything. I imagine it would all be in a separate file (e.g. `main.go`) so you could choose to override if if you wanted to, but did not have to.

The goal is definitely to make folks feel like they're writing (literally) a function, not building a web service. I'll make that explicit.